drivers/espi/espi_handlers.c

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright (c) 2020 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <drivers/espi.h>
+#include <syscall_handler.h>
+
+
+static inline int z_vrfy_espi_config(struct device *dev, struct espi_cfg *cfg)
+{
+       Z_OOPS(Z_SYSCALL_DRIVER_ESPI(dev, config));
+       Z_OOPS(Z_SYSCALL_MEMORY_READ(cfg, sizeof(*cfg)));
+
+       return z_impl_espi_config(dev, cfg);
+}
+#include <syscalls/espi_config_mrsh.c>
+
+static inline bool z_vrfy_espi_get_channel_status(struct device *dev,
+						  enum espi_channel ch)
+{
+       Z_OOPS(Z_SYSCALL_DRIVER_ESPI(dev, get_channel_status));
+
+       return z_impl_espi_get_channel_status(dev, ch);
+}
+#include <syscalls/espi_get_channel_status_mrsh.c>
+
+static inline int z_vrfy_espi_read_lpc_request(struct device *dev,
+					       enum lpc_peripheral_opcode op,
+					       u32_t *data)
+{
+       Z_OOPS(Z_SYSCALL_DRIVER_ESPI(dev, read_lpc_request));
+       Z_OOPS(Z_SYSCALL_MEMORY_WRITE(data, sizeof(*data)));
+
+       return z_impl_espi_read_lpc_request(dev, op, data);
```

Comment:
@Ceolin It is Friday and I would like to ask a dump question.  Function `espi_read_lpc_request` is has the out parameter `data`.  I've started looking at a few syscalls in linux (https://elixir.bootlin.com/linux/latest/source/kernel/sys.c) and for out parameters each syscall call first creates a temp variable in their execution context, then perform the underlying function call. Afterwards, the information from the temp variable is passed to the client using copy_to_user.  Shouldn't we be doing the same for out parameters? By the way I really appreciate the time you guys are investing on this !

Reply:
Hey, that is not dumb at all :) Andrew asked me to fix it in this PR.

Creating a temporary variable we protects against TOCTOU. For example, we have a struct with a buffer, during the verification we check if the caller has access to the struct memory and to the buffer, but if we don't make copies, it is possible to other thread with access to that struct change the buffer's address to point to something that the caller has no access but as the copy will occur in kernel mode this will succeed .

Reply:
> Hey, that is not dumb at all :) Andrew asked me to fix it in this PR.

Well, I had noted some issues with input parameters, but I think there is an issue with syscalls that have out parameters, and not just with this PR.

For syscalls that have output memory buffers, a lot of them currently do Z_SYSCALL_MEMORY_WRITE() on an out parameter before invoking the implementation function, and pass the out pointer directly to it.

This verifies that the caller has access to this memory, but may be unsafe. It should be safe if the output parameter is truly only used for output by the implementation. That is to say, this output buffer is ONLY written to, never read.

But if the system call uses this memory for internal working state, it's possible that there are preemption or SMP scenarios where this working state could get corrupted by some thread that has access to this memory, causing undefined behavior which at worst case could crash the kernel.

In addition, to properly implement in the future security features like SMAP, the kernel should never touch user memory at all except in designated copy_to_user()/copy_from_user() functions.

@ceolin I think we need to audit our system calls that have out parameters and change verifiers that look like this:

```
int z_vrfy_some_syscall(int *out_param)
{
    Z_OOPS(Z_SYSCALL_MEMORY_WRITE(out_param));
    return z_impl_some_syscall(out_param);
}
```
to
```
int z_vrfy_some_syscall(int *out_param)
{
    int local_out_param;
    int ret;
  
    ret = z_impl_some_syscall(&local_out_param);
	Z_OOPS(z_user_to_copy(out_param, &local_out_param, sizeof(*out_param)));
    return ret;
}
```
This might be a lot of syscalls that need to be fixed. I'll open a GH issue.

Reply:
https://github.com/zephyrproject-rtos/zephyr/issues/23877